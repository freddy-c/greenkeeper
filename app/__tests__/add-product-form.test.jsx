import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import renderer from "react-test-renderer";
import AddProductForm from "../product/add-product-form";
import SelectEvent from "react-select-event";
import { prismaMock } from "../singleton";

const mockManufacturers = [
    {
        id: "5b2aec59-6823-410d-b595-46b770584c56",
        name: "Aquatrols",
    },
    {
        id: "b6279c70-a9f6-49b9-85fb-5b01bdab7229",
        name: "Barenbrug",
    },
    {
        id: "e798c4f9-5e96-4bd9-be8c-8eeb126e711c",
        name: "ICL",
    },
];

describe("AddProductForm", () => {
    test("renders correctly", () => {
        const tree = renderer.create(<AddProductForm />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test("shows error message when form is submitted with invalid data", async () => {
        render(<AddProductForm manufacturers={mockManufacturers} />);

        fireEvent.click(screen.getByRole("button", { name: /add product/i }));

        await waitFor(() => {
            expect(screen.getAllByText(/Required/i)).toHaveLength(2);
        });
    });

    test("successfully submits form with valid data", async () => {
        const mockData = {
            id: "1",
            name: "New Product",
            manufacturerId: "b6279c70-a9f6-49b9-85fb-5b01bdab7229",
        };

        prismaMock.product.create.mockResolvedValue(mockData);

        render(<AddProductForm manufacturers={mockManufacturers} />);

        fireEvent.change(screen.getByLabelText("name"), {
            target: { value: "New Product" },
        });

        await SelectEvent.select(
            screen.getByLabelText(/manufacturer/i),
            "Barenbrug"
        );

        fireEvent.click(screen.getByRole("button", { name: /add product/i }));

        await waitFor(() => {
            expect(prismaMock.product.create).toHaveBeenCalledWith({
                data: {
                    name: mockData.name,
                    manufacturer: {
                        connect: {
                            id: mockData.manufacturerId,
                        },
                    },
                },
            });
        });
    });

    // test("reset form after successful submission", async () => {
    //     const mockData = {
    //         id: "1",
    //         name: "New Product",
    //         manufacturerId: "b6279c70-a9f6-49b9-85fb-5b01bdab7229",
    //     };

    //     prismaMock.product.create.mockResolvedValue(mockData);

    //     render(<AddProductForm manufacturers={mockManufacturers} />);

    //     fireEvent.change(screen.getByLabelText("name"), {
    //         target: { value: "New Product" },
    //     });

    //     await SelectEvent.select(
    //         screen.getByLabelText(/manufacturer/i),
    //         "Barenbrug"
    //     );

    //     fireEvent.click(screen.getByRole("button", { name: /add product/i }));

    //     await waitFor(() => {
    //         expect(screen.getByLabelText("name").value).toBe("");
    //         expect(screen.getByLabelText(/manufacturer/i).value).toBe("");
    //     });
    // });

    // test("handles errors during product creation", async () => {
    //     prismaMock.product.create.mockRejectedValue(
    //         new Error("Creation failed")
    //     );

    //     render(<AddProductForm manufacturers={mockManufacturers} />);

    //     fireEvent.change(screen.getByLabelText("name"), {
    //         target: { value: "New Product" },
    //     });

    //     await SelectEvent.select(
    //         screen.getByLabelText(/manufacturer/i),
    //         "Barenbrug"
    //     );

    //     fireEvent.click(screen.getByRole("button", { name: /add product/i }));

    //     await waitFor(() => {
    //         expect(prismaMock.product.create).toHaveBeenCalled();
    //         expect(
    //             screen.getByText(/Failed to create product/i)
    //         ).toBeInTheDocument();
    //     });
    // });
});
