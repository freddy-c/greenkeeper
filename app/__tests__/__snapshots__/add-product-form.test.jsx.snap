// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProductForm renders correctly 1`] = `
<div
  className="space-y-6 max-w-lg mx-auto"
>
  <form
    className="space-y-8"
    onSubmit={[Function]}
  >
    <div
      className="space-y-2"
    >
      <label
        className="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        htmlFor=":r0:-form-item"
        onMouseDown={[Function]}
      >
        Name
      </label>
      <input
        aria-describedby=":r0:-form-item-description"
        aria-invalid={false}
        aria-label="name"
        className="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id=":r0:-form-item"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder=""
      />
      <p
        className="text-sm text-muted-foreground"
        id=":r0:-form-item-description"
      >
        This is the name of the product.
      </p>
    </div>
    <div
      className="space-y-2"
    >
      <label
        className="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        htmlFor=":r1:-form-item"
        onMouseDown={[Function]}
      >
        Manufacturer
      </label>
      <div
        className="w-full css-b62m3t-container"
        id=":r1:-form-item"
        onKeyDown={[Function]}
      >
        <span
          className="css-1f43avz-a11yText-A11yText"
          id="react-select-2-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          className="css-1f43avz-a11yText-A11yText"
          role="log"
        />
        <div
          className=" css-1kup72l-control"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className=" css-1fdsijx-ValueContainer"
          >
            <div
              className=" css-dfjav8-placeholder"
              id="react-select-2-placeholder"
            >
              Select a manufacturer
            </div>
            <div
              className=" css-qbdosj-Input"
              data-value=""
            >
              <input
                aria-activedescendant=""
                aria-autocomplete="list"
                aria-describedby="react-select-2-placeholder"
                aria-expanded={false}
                aria-haspopup={true}
                aria-invalid={false}
                aria-label="manufacturer"
                autoCapitalize="none"
                autoComplete="off"
                autoCorrect="off"
                className=""
                disabled={false}
                id="manufacturer"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                role="combobox"
                spellCheck="false"
                style={
                  {
                    "background": 0,
                    "border": 0,
                    "color": "inherit",
                    "font": "inherit",
                    "gridArea": "1 / 2",
                    "label": "input",
                    "margin": 0,
                    "minWidth": "2px",
                    "opacity": 1,
                    "outline": 0,
                    "padding": 0,
                    "width": "100%",
                  }
                }
                tabIndex={0}
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            className=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              className=" css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              className=" css-1xc3v61-indicatorContainer"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-tj5bde-Svg"
                focusable="false"
                height={20}
                viewBox="0 0 20 20"
                width={20}
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="manufacturer"
          type="hidden"
          value=""
        />
      </div>
      <p
        className="text-sm text-muted-foreground"
        id=":r1:-form-item-description"
      >
        This is the manufacturer of the product.
      </p>
    </div>
    <button
      className="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
      type="submit"
    >
      Add Product
    </button>
  </form>
</div>
`;
